### PR TITLE
Propulsion.SqlStreamStore implementation

### PR DESCRIPTION
**Pull request by scrwtp**
An implementation of a Propulsion Source using SqlStreamStore. Doesn't include a Sink as there are gaps in version information that Equinox.SqlStreamStore provides on write conflicts.

Includes an implementation of an sql checkpoint ledger (using Dapper) and an in-memory test ledger.

### CODE REVIEW

File: src/Propulsion.SqlStreamStore/LockedIngester.fs [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+                    with
+                    | exc ->
+                        logger.Error(exc, "Exception while running StreamReader loop")
+                        return! Async.Raise exc
+            }
+
+        interface IDisposable with
+            member this.Dispose() =
+                work.CompleteAdding()
+                work.Dispose()
```
Comment (scrwtp) [inline src/Propulsion.SqlStreamStore/LockedIngester.fs:229]: Not really feeling strong on how a blocking collection should be disposed in a clean way - any suggestions?

File: src/Propulsion.SqlStreamStore/SqlStreamStoreSource.fs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+open StreamReader
+open LockedIngester
+
+type SqlStreamStoreSource =
+
+    static member Run
+            (logger : Serilog.ILogger,
+             store: IStreamStore,
+             ledger: ILedger,
```
Comment (scrwtp) [inline src/Propulsion.SqlStreamStore/SqlStreamStoreSource.fs:14]: Ledger implementation is pluggable as discussed before. Ledger.fs file has two implementations of it, SQL one and in-memory one.

File: src/Propulsion.SqlStreamStore/SqlStreamStoreSource.fs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+                             ?sleepInterval = readerSleepInterval,
+                             ?statsInterval = statsInterval)
+
+            let! position =
+                ledger.GetPosition { Stream = streamId; ConsumerGroup = consumerGroup }
+
+            let! ingesterComp = Async.StartChild (locked.Start())
+            let! readerComp = Async.StartChild (reader.Start(position))
+
+            do! Async.Parallel [ ingesterComp; readerComp ] |> Async.Ignore
```
Comment (scrwtp) [inline src/Propulsion.SqlStreamStore/SqlStreamStoreSource.fs:56]: Done this way so that errors thrown by ingester and reader respectively would propagate up to be handled by the client code - is this a reasonable approach or should it be done differently?

File: src/Propulsion.SqlStreamStore/SqlStreamStoreSource.fs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+            let boundedCapacity = defaultArg boundedCapacity 10
+
+            let logger =
+                let instanceId = System.Guid.NewGuid()
+                logger
+                    .ForContext("instanceId", string instanceId)
+                    .ForContext("consumerGroup", consumerGroup)
+
+            let ingester : Propulsion.Ingestion.Ingester<_,_> =
```
Comment (scrwtp) [inline src/Propulsion.SqlStreamStore/SqlStreamStoreSource.fs:35]: Should the ingester be explicitly stopped in a finally clause? I didn't see that done elsewhere in Propulsion and I'm not clear on who should manage this object's lifetime.

File: src/Propulsion.SqlStreamStore/LockedIngester.fs [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+                                    return b
+                            }
+
+                        stats.UpdateBatch(batch)
+
+                        logger.Debug("Trying to obtain the lock")
+
+                        // Try obtaining a distributed sql lock for the consumer group.
+                        // The lock would be released as part of commit operation on the underlying ingester.
+                        let! handle = ledger.Lock(consumerGroup)
```
Comment (scrwtp) [inline src/Propulsion.SqlStreamStore/LockedIngester.fs:181]: Locking is to allow multiple instances of a projection service running for a single consumer group for redundancy. I'm quite aware this is a bottle neck, but I'm not sure we can do any better without partitioning anyway?

Comment (bartelink) [reply src/Propulsion.SqlStreamStore/LockedIngester.fs:181]: For better or worse, Propulsion.EventStore assumes a single instance globally...
You can then use an orchestrator to manage the instance count; the writer is then expected to act idempotently in the face of multiple instances etc.

File: src/Propulsion.SqlStreamStore/SqlStreamStoreSource.fs [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+                             streamId,
+                             consumerGroup,
+                             ?maxBatchSize = readerMaxBatchSize,
+                             ?sleepInterval = readerSleepInterval,
+                             ?statsInterval = statsInterval,
+                             ?stopAfterInterval = readerStopAfterInterval)
+
+            try
+                let! position =
+                    checkpointer.GetPosition { Stream = streamId; ConsumerGroup = consumerGroup }
```
Comment (bartelink) [inline src/Propulsion.SqlStreamStore/SqlStreamStoreSource.fs:45]: replace record with an argument list?

File: src/Propulsion.SqlStreamStore/SqlStreamStoreSource.fs [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+                do! reader.Start(position)
+            finally
+                ingester.Stop()
+        }
+
+    /// Run SqlStreamStore.
+    static member Run
+        (logger : Serilog.ILogger,
+         store: IStreamStore,
```
Comment (bartelink) [inline src/Propulsion.SqlStreamStore/SqlStreamStoreSource.fs:55]: nits (most of these are in F# formatting guidelines, many in `.editorconfig`): 
- always have a CR at eof
- all decls should have spaces around `:`, i.e., `store : IStreamStore`
- commas in arg lists, tuple creations etc should have spaces before
- `spaces after `;`, e.g. `[<NoComparison; NoEquality>]`
- no space before method args (vs function args which always have space before), e.g. `logger.Information("message")`
- naming of args in arg lists should not have surrounding spaces, i.e. `?readerMaxBatchSize=readerMaxBatchSize`
- use `[]` in preference to `array`

File: src/Propulsion.SqlStreamStore/Propulsion.SqlStreamStore.fsproj [hunk 8]
```diff
@@ -0,0 +1,38 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+    <PropertyGroup>
+        <TargetFrameworks>netstandard2.0;net461</TargetFrameworks>
```
Comment (bartelink) [inline src/Propulsion.SqlStreamStore/Propulsion.SqlStreamStore.fsproj:4]: if you prefer, we can remove the `net461` and only add in the unlikely event someone happens to need it? (and maybe consider Microsoft.Data.SqlClient?)

File: src/Propulsion.SqlStreamStore/Checkpoint.fs [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+            match Seq.tryHead res with
+            | Some res ->
+                return res.Position
+            | None ->
+                return Nullable()
+        }
+
+type SqlCheckpointer(connString : string) =
+
+    member this.CreateIfNotExists() =
```
Comment (bartelink) [inline src/Propulsion.SqlStreamStore/Checkpoint.fs:92]: I know CosmosDb uses this naming and its not too surprising, but perhaps adding Schema or Table in the name might make it more clear what the role of this is (admittedly the interface vs member difference makes it pretty clear)

File: src/Propulsion.SqlStreamStore/Infrastructure.fs [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+    let toNullable option =
+        match option with
+        | None -> System.Nullable()
+        | Some v -> System.Nullable(v)
+
+module private Seq =
+
+    let tryHead (source : seq<_>) =
+        if source = null then
```
Comment (bartelink) [inline src/Propulsion.SqlStreamStore/Infrastructure.fs:47]: can this check go into the mainline code so we can eventually slot in the one from FSharp.Core and not end up with breaking logic?

File: src/Propulsion.SqlStreamStore/Infrastructure.fs [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+                    if t.IsFaulted then
+                        let e = t.Exception
+                        if e.InnerExceptions.Count = 1 then ek e.InnerExceptions.[0]
+                        else ek e
+                    elif t.IsCanceled then ek (TaskCanceledException("Task wrapped with Async has been cancelled."))
+                    elif t.IsCompleted then k ()
+                    else ek(Exception "invalid Task state!"))
+                |> ignore
+
+//#if NET461
```
Comment (bartelink) [inline src/Propulsion.SqlStreamStore/Infrastructure.fs:36]: better if this guard works than be an ambiguous comment

File: src/Propulsion.SqlStreamStore/Locked.fs [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+open System
+open SqlStreamStore
+open Medallion.Threading.Sql
+
+module Locked =
+
+    /// Creates a distributed lock that manages its own db connection.
+    let distributedLock (connString: string) (consumerGroup: string) =
+        let lockName = SqlDistributedLock.GetSafeLockName(sprintf "lock-%s" consumerGroup)
+        SqlDistributedLock(lockName, connString, SqlDistributedLockConnectionStrategy.Azure)
```
Comment (bartelink) [inline src/Propulsion.SqlStreamStore/Locked.fs:12]: should the Azure go up into one of the names to make it clear that this is tuned for / only considers Azure?

File: src/Propulsion.SqlStreamStore/StreamReader.fs [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+                    msg.Type,
+                    len0ToNull data,
+                    len0ToNull metadata,
+                    msg.MessageId,
+                    timestamp = DateTimeOffset(msg.CreatedUtc)
+                )
+
+            { stream = StreamName.internalParseSafe msg.StreamId; event = event }
+
+    type Stats(logger: ILogger, ?statsInterval: TimeSpan) =
```
Comment (bartelink) [inline src/Propulsion.SqlStreamStore/StreamReader.fs:56]: newer code pushes this defaulting outward (i.e. it goes into a `type Arguments` in `dotnet-templates` )

File: src/Propulsion.SqlStreamStore/StreamReader.fs [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+        streamId,
+        consumerGroup,
+        ?maxBatchSize: int,
+        ?sleepInterval: TimeSpan,
+        ?statsInterval: TimeSpan,
+        ?stopAfterInterval: TimeSpan
+    ) =
+
+    let maxBatchSize = defaultArg maxBatchSize 100
+    let sleepInterval = defaultArg sleepInterval (TimeSpan.FromSeconds(5.))
```
Comment (bartelink) [inline src/Propulsion.SqlStreamStore/StreamReader.fs:134]: 5s is a really long default - maybe make it mandatory so caller makes it explicit?

File: src/Propulsion.SqlStreamStore/StreamReader.fs [hunk 15]
```diff
@@ -0,0 +1,226 @@
...
+    let stopAfterInterval = defaultArg stopAfterInterval TimeSpan.MaxValue
+
+    let stats = Stats(logger, ?statsInterval = statsInterval)
+
+    let commit batch =
+        async {
+            try
+                do! ledger.CommitPosition { Stream = streamId; ConsumerGroup = consumerGroup; Position = Nullable(batch.lastPosition) }
```
Comment (bartelink) [inline src/Propulsion.SqlStreamStore/StreamReader.fs:142]: remove term `ledger`

File: src/Propulsion.SqlStreamStore/StreamReader.fs [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+        }
+
+    let processPage (page: ReadAllPage) =
+        async {
+            if page.Messages.Length > 0 then
+
+                let events =
+                    page.Messages
+                    |> Seq.map StreamMessage.intoStreamEvent
+                    |> Seq.sortBy (fun x -> x.event.Index)
```
Comment (bartelink) [inline src/Propulsion.SqlStreamStore/StreamReader.fs:158]: why is this necessary?

File: src/Propulsion.SqlStreamStore/StreamReader.fs [hunk 17]
```diff
@@ -0,0 +1,226 @@
...
+
+                let batch =
+                    {
+                        firstPosition = events.[0].event.Index
+                        lastPosition = events.[events.Length - 1].event.Index
+                        messages = events
+                        isEnd = page.IsEnd
+                    }
+
+                logger.Debug("Submitting a batch of {batchSize} events, position {firstPosition} through {lastPosition}",
```
Comment (bartelink) [inline src/Propulsion.SqlStreamStore/StreamReader.fs:169]: those are not positions - the batch might cross multiple streams

File: src/Propulsion.SqlStreamStore/StreamReader.fs [hunk 18]
```diff
@@ -0,0 +1,226 @@
...
+                    do! processPage page
+
+                    // If processPage was successful, ask for new page on the next iteration
+                    workItem <- Work.TakeNext page
+
+                    if page.IsEnd then
+                        do! Async.Sleep sleepInterval
+                with
+                | exc ->
+                    logger.Error(exc, "Exception while running StreamReader loop")
```
Comment (bartelink) [inline src/Propulsion.SqlStreamStore/StreamReader.fs:222]: this is def an Error if you are going to kill the app - are you? maybe even `Fatal` ? maybe include text indicating you're abending?

File: src/Propulsion.SqlStreamStore/LockedIngester.fs (hunk 5 above)
Comment (bartelink) [reply src/Propulsion.SqlStreamStore/LockedIngester.fs]: (discussed out of band - the locking is not intrinsic to the functionality, the `proProjector` template will illustrate its use)

File: src/Propulsion.SqlStreamStore/Locked.fs (hunk 12 above)
Comment (scrwtp) [reply src/Propulsion.SqlStreamStore/Locked.fs]: I've actually pulled out the locking specific parts from the PR. It mostly addresses a redundancy problem that is not core to what Propulsion wants to achieve, so I don't feel it's worth having it here. Check the update.

File: src/Propulsion.SqlStreamStore/StreamReader.fs (hunk 14 above)
Comment (scrwtp) [reply src/Propulsion.SqlStreamStore/StreamReader.fs:131]: Yeah, it's long. Propose another default?

Comment (bartelink) [reply src/Propulsion.SqlStreamStore/StreamReader.fs:131]: I'd propose to remove defaulting and let the value live in Program.fs in a template
For Es, the TailInterval default I happen to use is -1s
(Also that name makes it clearer what the sleep is for and when it happens)

File: src/Propulsion.SqlStreamStore/StreamReader.fs (hunk 18 above)
Comment (scrwtp) [reply src/Propulsion.SqlStreamStore/StreamReader.fs]: Depends on the client - also moving that statement try with outside into SSSS.Run since I removed locking.

File: src/Propulsion.SqlStreamStore/StreamReader.fs (hunk 14 above)
Comment (bartelink) [reply src/Propulsion.SqlStreamStore/StreamReader.fs:131]: Chances are I derived the 1s from the default value for the CFP one: https://github.com/jet/propulsion/search?q=feedpolldelay&unscoped_q=feedpolldelay
(that, and the sqlstreamstore one, could both do with a comment as to what the default value is)

File: src/Propulsion.SqlStreamStore/Locked.fs (hunk 12 above)
Comment (bartelink) [reply src/Propulsion.SqlStreamStore/Locked.fs]: 👍 makes sense

File: src/Propulsion.SqlStreamStore/StreamReader.fs (hunk 18 above)
Comment (bartelink) [reply src/Propulsion.SqlStreamStore/StreamReader.fs]: I guess my point was that using Error level for something that might ultimately be handled by an outer layer might be annoying for someone else using the lib?

File: src/Propulsion.SqlStreamStore/Propulsion.SqlStreamStore.fsproj (hunk 8 above)
Comment (scrwtp) [reply src/Propulsion.SqlStreamStore/Propulsion.SqlStreamStore.fsproj:4]: 1. Your choice, I don't really see the need to target net461 anymore myself, just kept with what other projects were doing.
2. I'll drop SqlClient dependency from there, it's not a business of this project to restrict it, it's only there for Dapper which I believe supports both versions.

File: src/Propulsion.SqlStreamStore/Checkpoint.fs (hunk 9 above)
Comment (scrwtp) [reply src/Propulsion.SqlStreamStore/Checkpoint.fs:92]: I'll add *Schema* to the name.

File: src/Propulsion.SqlStreamStore/SqlStreamStoreSource.fs (hunk 6 above)
Comment (scrwtp) [reply src/Propulsion.SqlStreamStore/SqlStreamStoreSource.fs:43]: I went for a record to keep it in line with the other interface member, if I'm changing it, I'll change both members.

File: src/Propulsion.SqlStreamStore/Infrastructure.fs (hunk 10 above)
Comment (scrwtp) [reply src/Propulsion.SqlStreamStore/Infrastructure.fs:47]: why would the logic break?

File: src/Propulsion.SqlStreamStore/Infrastructure.fs (hunk 11 above)
Comment (scrwtp) [reply src/Propulsion.SqlStreamStore/Infrastructure.fs:36]: Haha, idk how that happened :D

File: src/Propulsion.SqlStreamStore/StreamReader.fs (hunk 17 above)
Comment (scrwtp) [reply src/Propulsion.SqlStreamStore/StreamReader.fs:165]: In all cases where I'm using the term position, I'm refering to the position in the (non-existent) $all stream.

File: src/Propulsion.SqlStreamStore/StreamReader.fs (hunk 13 above)
Comment (scrwtp) [reply src/Propulsion.SqlStreamStore/StreamReader.fs:54]: so what specifically do you prefer here? Make statsInterval non-optional?

File: src/Propulsion.SqlStreamStore/StreamReader.fs (hunk 18 above)
Comment (scrwtp) [reply src/Propulsion.SqlStreamStore/StreamReader.fs]: But you said Fatal? :D Warn is acceptable?

File: src/Propulsion.SqlStreamStore/StreamReader.fs (hunk 16 above)
Comment (scrwtp) [reply src/Propulsion.SqlStreamStore/StreamReader.fs:154]: Just to make sure the logs come out right, I don't assume page.Messages are ordered, even if they probably are.

File: src/Propulsion.SqlStreamStore/Propulsion.SqlStreamStore.fsproj (hunk 8 above)
Comment (bartelink) [reply src/Propulsion.SqlStreamStore/Propulsion.SqlStreamStore.fsproj:4]: Its definitely not a prerequisite, and if it involves conditionals and/or extra helpers, that's definitely a reason to consider removing it.
The blanket support for net461 made a lot more sense 2 years ago and for libraries that would be referenced from an app's "core" modules - realistically anything using `Propulsion.SqlStreamStore` should be a small new separated console app.

File: src/Propulsion.SqlStreamStore/Infrastructure.fs (hunk 10 above)
Comment (bartelink) [reply src/Propulsion.SqlStreamStore/Infrastructure.fs:47]: I would expect that stock `Sql.tryHead` would follow `Enumerable.FirstOrDefault` and do a `nullArg` if you called it on a `seq` that's `null` ?

File: src/Propulsion.SqlStreamStore/StreamReader.fs (hunk 13 above)
Comment (bartelink) [reply src/Propulsion.SqlStreamStore/StreamReader.fs:54]: Yes, non-optional (I pushed a lot of these out about a month ago as there are better ways to make the consumer logic succinct than having defaults inside Propulsion itself)

File: src/Propulsion.SqlStreamStore/StreamReader.fs (hunk 17 above)
Comment (bartelink) [reply src/Propulsion.SqlStreamStore/StreamReader.fs:165]: ok, but those values are derived from the `Index` value on the event - those values are the ordinal number of the event within the stream
If that's not the case, we have a different problem, as Propulsion expects `Index` values to be contiguous within a Stream

File: src/Propulsion.SqlStreamStore/SqlStreamStoreSource.fs (hunk 6 above)
Comment (bartelink) [reply src/Propulsion.SqlStreamStore/SqlStreamStoreSource.fs:43]: 👍 In general, using records in this way is not the convention in Propulsion

File: src/Propulsion.SqlStreamStore/StreamReader.fs (hunk 18 above)
Comment (bartelink) [reply src/Propulsion.SqlStreamStore/StreamReader.fs]: I am not telling you what way to fall on it - its unclear to me. Saying we need to pick from
1. lib is pushing it out to the caller to own what happens overall (then it should be a warning and the outher can take the real responsibility, with appropriate extra logging as to the conclusion)
2. lib is taking an opinionated stance - there's nothing we can do other than force abend (then it would be `Fatal`)

File: src/Propulsion.SqlStreamStore/Infrastructure.fs (hunk 10 above)
Comment (scrwtp) [reply src/Propulsion.SqlStreamStore/Infrastructure.fs:47]: Oh, so that's why it's implemented in such a deranged way.

I'll drop the null check but it's somewhat of a moot point anyway.

File: src/Propulsion.SqlStreamStore/Propulsion.SqlStreamStore.fsproj (hunk 8 above)
Comment (scrwtp) [reply src/Propulsion.SqlStreamStore/Propulsion.SqlStreamStore.fsproj:4]: Ok, if it's not necessary, I'll drop support for it.

File: src/Propulsion.SqlStreamStore/StreamReader.fs (hunk 16 above)
Comment (bartelink) [reply src/Propulsion.SqlStreamStore/StreamReader.fs:154]: I believe based on what I know of SSS (and how it works in ES) that they are guaranteed to be in order of Position
However I believe that (see other comment re conflating Index and Position) this would end up ordering them by ordinal in the stream, which would not make sense

File: src/Propulsion.SqlStreamStore/StreamReader.fs (hunk 17 above)
Comment (scrwtp) [reply src/Propulsion.SqlStreamStore/StreamReader.fs:165]: And what happens if they aren't? 

It's a simple change in the event mapping to source it from the stream instead of global position.

File: src/Propulsion.SqlStreamStore/StreamReader.fs [hunk 19]
```diff
@@ -0,0 +1,226 @@
...
+            let data =
+                let json = msg.GetJsonData().Result
+                Encoding.UTF8.GetBytes(json)
+
+            let metadata =
+                Encoding.UTF8.GetBytes(msg.JsonMetadata)
+
+            let event =
+                FsCodec.Core.TimelineEvent.Create (
+                    msg.Position,
```
Comment (bartelink) [inline src/Propulsion.SqlStreamStore/StreamReader.fs:44]: hm, missed this - see other comments
this needs to be the 0-based Index of the event within its original/actual stream, i.e. the `StreamId`
The Index is used to deduplicate/group/order events per stream when buffering
The Position itself is only used for checkpointing purposes (if consuming from CFP or Kafka, a Position is an aggregate value)
(If there was a need to propagate this context inward to a Handler, the TimelineEvent's `Context` property can be used to carry such information, but that has not been done elsewhere)

File: src/Propulsion.SqlStreamStore/Infrastructure.fs (hunk 10 above)
Comment (bartelink) [reply src/Propulsion.SqlStreamStore/Infrastructure.fs:47]: >Oh, so that's why it's implemented in such a deranged way.

My 9yo normally prefixes such commentary with a "No offence" ;)

Yeah, I reckon removing is for the best ;)

File: src/Propulsion.SqlStreamStore/StreamReader.fs (hunk 19 above)
Comment (scrwtp) [reply src/Propulsion.SqlStreamStore/StreamReader.fs:44]: Good catch, changing that should not be a problem.

Comment (scrwtp) [reply src/Propulsion.SqlStreamStore/StreamReader.fs:44]: But as far as deduplication and ordering goes, global position should work just fine anyway? Unless it actually is looking for gaps in the indices.

File: src/Propulsion.SqlStreamStore/StreamReader.fs (hunk 17 above)
Comment (bartelink) [reply src/Propulsion.SqlStreamStore/StreamReader.fs:165]: Position is what's needed for logging in the context of this reader - what you have makes sense overall
But see https://github.com/jet/propulsion/pull/72#discussion_r435193099 for more detail on what purpose/reqs `Index` serves

File: src/Propulsion.SqlStreamStore/StreamReader.fs (hunk 19 above)
Comment (bartelink) [reply src/Propulsion.SqlStreamStore/StreamReader.fs:44]: The idea is that the StreamsScheduler's buffer holds stuff as arrays per stream in order to deal with incomplete replays

e.g. if kafka input has (for a given stream) 1,2,3 ... 1, 2 ... 3,4 

then the handler is called with contiguous spans with duplicates dropped
so it will typically. be called with [1,2,3] or [1,2,3,4]

I _think_ you're correct that it would actually "work", but each handler invocation would be with a single event only (and anyone depending on Index to mean what it's supposed to mean (ordinal number within the stream) would have issues)

File: src/Propulsion.SqlStreamStore/StreamReader.fs (hunk 16 above)
Comment (scrwtp) [reply src/Propulsion.SqlStreamStore/StreamReader.fs:154]: Yeah, that needs to change or go away when the index/position change happens.

File: src/Propulsion.SqlStreamStore/SqlStreamStoreSource.fs [hunk 20]
```diff
@@ -3,17 +3,28 @@ namespace Propulsion.SqlStreamStore
 open System
 open SqlStreamStore
 
+type ReaderSpec =
```
Comment (bartelink) [inline src/Propulsion.SqlStreamStore/SqlStreamStoreSource.fs:6]: Hm not sure I love the existing ReaderSpec - that came about as there are so many knobs in the ES impl
Also, the global default block with overrides pattern is better replaced with individual commented optional args elsewhere 
though if you want to argue symmetry with .ES has value I guess thats fair...

File: src/Propulsion.SqlStreamStore/StreamReader.fs [hunk 21]
```diff
@@ -0,0 +1,207 @@
...
+
+    let commit batch =
+        async {
+            try
+                do! checkpointer.CommitPosition(streamId, consumerGroup, batch.lastPosition)
+                stats.UpdateCommitedPosition(batch.lastPosition)
+                logger.Debug("Committed position {position}", batch.lastPosition)
+            with
+            | exc ->
+                logger.Error(exc, "Exception while commiting position {position}", batch.lastPosition)
```
Comment (bartelink) [inline src/Propulsion.SqlStreamStore/StreamReader.fs:138]: should this be a Warning as it is being rethrown?

File: src/Propulsion.SqlStreamStore/StreamReader.fs [hunk 22]
```diff
@@ -0,0 +1,207 @@
...
+        streamId,
+        consumerGroup,
+        maxBatchSize: int,
+        tailSleepInterval: TimeSpan,
+        statsInterval: TimeSpan
+    ) =
+
+    let stats = Stats(logger, statsInterval)
+
+    let commit batch =
```
Comment (bartelink) [inline src/Propulsion.SqlStreamStore/StreamReader.fs:130]: only lastPosition is used from `batch` ?

**Comment by bartelink** [comment]
😎 All looking very lean and neat now

File: src/Propulsion.SqlStreamStore/StreamReader.fs (hunk 21 above)
Comment (scrwtp) [reply src/Propulsion.SqlStreamStore/StreamReader.fs:138]: :+1: Will change

File: src/Propulsion.SqlStreamStore/SqlStreamStoreSource.fs (hunk 20 above)
Comment (scrwtp) [reply src/Propulsion.SqlStreamStore/SqlStreamStoreSource.fs:6]: For me this many knobs without defaults was already too many. Symmetry with ES is a bonus.

File: src/Propulsion.SqlStreamStore/StreamReader.fs (hunk 22 above)
Comment (scrwtp) [reply src/Propulsion.SqlStreamStore/StreamReader.fs:130]: As in, just pass position here? Can be done.

File: src/Propulsion.SqlStreamStore/SqlStreamStoreSource.fs (hunk 20 above)
Comment (bartelink) [reply src/Propulsion.SqlStreamStore/SqlStreamStoreSource.fs:6]: Yes, your call - however FsKafka has shown you can represent N options that have defaults as N optional args to a `Start` or `Create` that default.

The main thing you lose is the ability to couple all the settings in a record with minimal docs and pass them from one end of the app to another without thinking.

In general, this is a 'pattern' I'm looking to de-emphasize

File: src/Propulsion.SqlStreamStore/StreamReader.fs (hunk 22 above)
Comment (bartelink) [reply src/Propulsion.SqlStreamStore/StreamReader.fs:130]: 👍

File: src/Propulsion.SqlStreamStore/SqlStreamStoreSource.fs (hunk 7 above)
Comment (scrwtp) [reply src/Propulsion.SqlStreamStore/SqlStreamStoreSource.fs]: I made some changes here, but beyond that feel free to run whatever linter/formatter you use ;)

Comment (bartelink) [reply src/Propulsion.SqlStreamStore/SqlStreamStoreSource.fs]: Thanks, as mention it all looks pretty consistent now (as it happens I've realised only today that this repo's spacing around colons is not in accordance with https://docs.microsoft.com/en-us/dotnet/fsharp/style-guide/formatting - I for one welcome our incoming final linter overlord!)

**Comment by bartelink** [comment]
Thanks for bearing with me across all the to/fro - looks great now
As discussed I'll merge and release this shortly
